pkg/cloud/services/compute/instance.go

```diff
@@ -99,36 +97,51 @@ func (s *Service) InstanceCreate(clusterName string, machine *clusterv1.Machine,
 	}
 	// Get all network UUIDs
 	var nets []ServerNetwork
-	for _, net := range openStackMachine.Spec.Networks {
-		opts := networks.ListOpts(net.Filter)
-		opts.ID = net.UUID
-		ids, err := getNetworkIDsByFilter(s.networkClient, &opts)
-		if err != nil {
-			return nil, err
-		}
-		for _, netID := range ids {
-			if net.Subnets == nil {
-				nets = append(nets, ServerNetwork{
-					networkID: netID,
-				})
+	if len(openStackMachine.Spec.Networks) > 0 {
+		for _, net := range openStackMachine.Spec.Networks {
+			opts := networks.ListOpts(net.Filter)
+			opts.ID = net.UUID
+			ids, err := networking.GetNetworkIDsByFilter(s.networkClient, &opts)
+			if err != nil {
+				return nil, err
 			}
-
-			for _, subnet := range net.Subnets {
-				subnetOpts := subnets.ListOpts(subnet.Filter)
-				subnetOpts.ID = subnet.UUID
-				subnetOpts.NetworkID = netID
-				subnetsByFilter, err := networking.GetSubnetsByFilter(s.networkClient, &subnetOpts)
-				if err != nil {
-					return nil, err
-				}
-				for _, subnetByFilter := range subnetsByFilter {
+			for _, netID := range ids {
+				if net.Subnets == nil {
 					nets = append(nets, ServerNetwork{
-						networkID: subnetByFilter.NetworkID,
-						subnetID:  subnetByFilter.ID,
+						networkID: netID,
 					})
+					continue
+				}
+
+				for _, subnet := range net.Subnets {
+					subnetOpts := subnets.ListOpts(subnet.Filter)
+					subnetOpts.ID = subnet.UUID
+					subnetOpts.NetworkID = netID
+					subnetsByFilter, err := networking.GetSubnetsByFilter(s.networkClient, &subnetOpts)
+					if err != nil {
+						return nil, err
+					}
+					for _, subnetByFilter := range subnetsByFilter {
+						nets = append(nets, ServerNetwork{
+							networkID: subnetByFilter.NetworkID,
+							subnetID:  subnetByFilter.ID,
+						})
+					}
 				}
 			}
 		}
+	} else {
+		if openStackCluster.Status.Network == nil {
+			return nil, fmt.Errorf("no network was found in OpenStackCluster.Status.Network")
```

Comment:
maybe we need add some info that because the machine doesn't have network spec and cluster network is also nil
so we return error..

Reply:
Agree. Fixed